### PR TITLE
Feature/dynbindecistree

### PR DESCRIPTION
**Pull request by jinseisieko**
- Implementation of dynamic decision tree
- Template for docs

### CODE REVIEW

File: docs/design/roadmap.md [hunk 1]
```diff
@@ -0,0 +1,291 @@
+# Development Roadmap
```
Comment (jinseisieko) [inline docs/design/roadmap.md:1]: Add issues with some of enhancements

File: docs/design/enhancements.md [hunk 2]
```diff
@@ -0,0 +1,491 @@
+# Proposed Enhancements
```
Comment (jinseisieko) [inline docs/design/enhancements.md:1]: Add issues with some of enhancements

File: src/main/java/io/github/jinseisieko/decisiontree/ArrayDecisionTree.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+    @Override
+    public boolean isFullyDefined() {
+        throw new UnsupportedOperationException("Not supported yet.");
+    }
+
+    public class Builder implements DecisionTreeBuilder<I, O> {
+        private ArrayDecisionTree<I, O> tree;
+        private int depth;
+        private Predicate<I>[] nodes;
```
Comment (jinseisieko) [inline src/main/java/io/github/jinseisieko/decisiontree/ArrayDecisionTree.java]: Isn't it redundant

File: src/main/java/io/github/jinseisieko/decisiontree/ConditionNode.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
+package io.github.jinseisieko.decisiontree;
+
+import java.util.Objects;
+import java.util.function.Predicate;
+
+final class ConditionNode<I, O> extends AbstractDecisionNode<I, O> {
+    private Predicate<I> condition;
+
+    public ConditionNode(Predicate<I> condition) {
+        this.setCondition(condition);
```
Comment (jinseisieko) [inline src/main/java/io/github/jinseisieko/decisiontree/ConditionNode.java:10]: Objects.requireNonNull(condition, "Condition cannot be null");

File: src/main/java/io/github/jinseisieko/decisiontree/DynamicDecisionBranch.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        }
+        AbstractDecisionNode<I, O> falseChild = rootNode.getFalseChild();
+        if (falseChild == null) {
+            throw new IllegalStateException("Cannot get false branch: false child does not exist");
+        }
+        return new DynamicDecisionBranch<>(builder, falseChild);
+    }
+
+    @Override
+    public int getDepth() {
```
Comment (jinseisieko) [inline src/main/java/io/github/jinseisieko/decisiontree/DynamicDecisionBranch.java:41]: It can be optimized via maintaining the depth

File: src/main/java/io/github/jinseisieko/decisiontree/ConditionNode.java (hunk 4 above)
Comment (jinseisieko) [reply src/main/java/io/github/jinseisieko/decisiontree/ConditionNode.java:10]: It is implemented in setCondition

Comment (jinseisieko) [reply src/main/java/io/github/jinseisieko/decisiontree/ConditionNode.java:10]: It is redundant because the class is final

File: src/main/java/io/github/jinseisieko/decisiontree/DynamicDecisionTree.java [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+                } else if (parent.getTrueChild() == currentNode) {
+                    parent.setTrueChild(newNode);
+                } else {
+                    parent.setFalseChild(newNode);
+                }
+                currentNode = newNode;
+            }
+        }
+
+        private AbstractDecisionNode<I, O> findParent(AbstractDecisionNode<I, O> root,
```
Comment (jinseisieko) [inline src/main/java/io/github/jinseisieko/decisiontree/DynamicDecisionTree.java:143]: recreate logic of this approach, store parent into node

File: src/main/java/io/github/jinseisieko/decisiontree/DynamicDecisionTree.java [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+        @Override
+        public DecisionTreeBuilder<I, O> goToTrueChild() {
+            if (currentNode == null) {
+                throw new IllegalStateException("Cannot navigate: tree is empty");
+            }
+            if (!(currentNode instanceof ConditionNode)) {
+                throw new IllegalStateException("Cannot navigate: current node is not a condition node");
+            }
+            AbstractDecisionNode<I, O> trueChild = currentNode.getTrueChild();
+            if (trueChild == null) {
```
Comment (jinseisieko) [inline src/main/java/io/github/jinseisieko/decisiontree/DynamicDecisionTree.java:167]: Is it good approach, is it single ?

File: src/main/java/io/github/jinseisieko/decisiontree/DynamicDecisionTree.java [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+                return this;
+            }
+            if (currentNode instanceof ConditionNode) {
+                currentNode.clearChildren();
+            }
+            return this;
+        }
+
+        @Override
+        public DecisionTreeBuilder<I, O> insertBranch(DecisionBranch<I, O> branch) {
```
Comment (jinseisieko) [inline src/main/java/io/github/jinseisieko/decisiontree/DynamicDecisionTree.java:267]: Possibly there should be copy of branch instead of linkage assignment

File: src/main/java/io/github/jinseisieko/decisiontree/DynamicDecisionTree.java [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+            return new DynamicDecisionBranch<>(this, currentNode);
+        }
+
+        @Override
+        public DecisionBranch<I, O> asBranchOfSubtree() {
+            if (currentNode == null) {
+                throw new IllegalStateException("Cannot create branch: current node is null");
+            }
+            AbstractDecisionNode<I, O> copiedNode = currentNode.copy();
+            return new DynamicDecisionBranch<>(new Builder(tree, copiedNode), copiedNode);
```
Comment (jinseisieko) [inline src/main/java/io/github/jinseisieko/decisiontree/DynamicDecisionTree.java:305]: wrong logic of creating subtree

File: src/main/java/io/github/jinseisieko/decisiontree/DynamicDecisionTree.java [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+
+        @Override
+        public int getDepth() {
+            if (tree.root == null) {
+                throw new IllegalStateException("Cannot get depth: tree is empty");
+            }
+            return calculateDepth(tree.root);
+        }
+
+        private int calculateDepth(AbstractDecisionNode<I, O> node) {
```
Comment (jinseisieko) [inline src/main/java/io/github/jinseisieko/decisiontree/DynamicDecisionTree.java:366]: It can be optimized via maintaining the depth of tree

File: src/main/java/io/github/jinseisieko/decisiontree/DynamicDecisionTree.java [hunk 11]
```diff
@@ -0,0 +1,430 @@
...
+
+        private boolean isNodeComplete(AbstractDecisionNode<I, O> node) {
+            if (node == null) {
+                return false;
+            }
+            return node.isComplete();
+        }
+
+        @Override
+        public boolean isFullyDefined() {
```
Comment (jinseisieko) [inline src/main/java/io/github/jinseisieko/decisiontree/DynamicDecisionTree.java:394]: Can be maintained

File: src/test/java/io/github/jinseisieko/decisiontree/ConditionNodeTest.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import org.junit.jupiter.api.Test;
+
+import static io.github.jinseisieko.decisiontree.TestUtilities.alwaysTrue;
+import static io.github.jinseisieko.decisiontree.TestUtilities.assertThrowsWithNonEmptyMessage;
+
+class ConditionNodeTest {
+
```
Comment (jinseisieko) [inline src/test/java/io/github/jinseisieko/decisiontree/ConditionNodeTest.java:11]: Add test with constructor with null should throw the nullPointerEx

File: src/main/java/io/github/jinseisieko/decisiontree/DynamicDecisionTree.java [hunk 13]
```diff
@@ -0,0 +1,430 @@
...
+        public DecisionTreeBuilder<I, O> goToTrueChild() {
+            if (currentNode == null) {
+                throw new IllegalStateException("Cannot navigate: tree is empty");
+            }
+            if (!(currentNode instanceof ConditionNode)) {
+                throw new IllegalStateException("Cannot navigate: current node is not a condition node");
+            }
+            AbstractDecisionNode<I, O> trueChild = currentNode.getTrueChild();
+            if (trueChild == null) {
+                trueChild = new OutcomeNode<>(input -> {
```
Comment (jinseisieko) [inline src/main/java/io/github/jinseisieko/decisiontree/DynamicDecisionTree.java:168]: Add constructor with this implementation as default

File: src/main/java/io/github/jinseisieko/decisiontree/OutcomeNode.java [hunk 14]
```diff
@@ -0,0 +1,63 @@
+package io.github.jinseisieko.decisiontree;
+
+import java.util.Objects;
+import java.util.function.Function;
+
+final class OutcomeNode<I, O> extends AbstractDecisionNode<I, O> {
```
Comment (jinseisieko) [inline src/main/java/io/github/jinseisieko/decisiontree/OutcomeNode.java:6]: Add default constructor with implementation that throws exception of incomplete

File: src/main/java/io/github/jinseisieko/decisiontree/DynamicDecisionTree.java (hunk 11 above)
Comment (jinseisieko) [reply src/main/java/io/github/jinseisieko/decisiontree/DynamicDecisionTree.java:394]: not important

File: src/main/java/io/github/jinseisieko/decisiontree/DynamicDecisionTree.java (hunk 9 above)
Comment (jinseisieko) [reply src/main/java/io/github/jinseisieko/decisiontree/DynamicDecisionTree.java:305]: I was dumb

File: src/main/java/io/github/jinseisieko/decisiontree/OutcomeNode.java (hunk 14 above)
Comment (jinseisieko) [reply src/main/java/io/github/jinseisieko/decisiontree/OutcomeNode.java:6]: Issue is cerated

File: src/test/java/io/github/jinseisieko/decisiontree/ConditionNodeTest.java (hunk 12 above)
Comment (jinseisieko) [reply src/test/java/io/github/jinseisieko/decisiontree/ConditionNodeTest.java:11]: fixed
